[WIP] 2D map generator tool: wallmap

arl opened the pull request:
Add `wallmap` source, go executable to rasterize the walls `.obj` file into a bitmap.
- `objloader.go`: super basic obj GO file loader, that spits triangles
- `wallmap.go`: cli executable, takes a `-scale` param and generates the bitmap.

Image size and origin offset:
- If `maxx`, `maxy` are the highest `x` and `y` values of the mesh vertices, then the resulting bitmap will have a width of `scale * maxx` and a height of `scale * maxy`.  
- by convention, pixel coordinates of the 2D bitmap **do not need to be translated** in order to be converted to _Game Units_, in other words the point `0,0` of the world **IS** the origin of the bitmap. The only transformation needed to pass from _Game Coord_ to _Bitmap Coord_ is a scaling, the one that was defined when running `wallmap`. That is why there should not be any negative `X` or `Y` coordinates in the terrain OBJ file as they would imply an additional translation. The tool should refuse to generate the bitmap **in case of negative coordinates**, or at the very least emit a warning.

**NOTE:** source code is added to `ROOT/src/tools/` because it has to be in the `$GOPATH`
Maybe `ROOT/src/tool` could become the new tool destination, for this reason.

**TODO**:
- Add a README with an example of use

arl committed:
First working version of wallmap: 2D map generator

arl committed:
Add basic go get deps script

rasky commented on src/tools/wallmap/objloader.go:
```diff
@@ -0,0 +1,194 @@
+}
+
+func (p *Point3) Set(s []string) error {
+	if _, err := fmt.Sscanf(s[0], "%f", &p.X); err != nil {
```
`strconv.ParseFloat` is much faster; also `Sscanf` is always surprising with whitespace.

rasky commented on src/tools/wallmap/objloader.go:
```diff
@@ -0,0 +1,194 @@
+	p.Y = p.Y * f
+}
+
+func Point2From3(p Point3) Point2 {
```
This is technically a constructor, so I would call it `NewPoint2From3`

rasky commented on src/tools/wallmap/objloader.go:
```diff
@@ -0,0 +1,194 @@
+	area := ((t.P2.X-t.P1.X)*(t.P3.Y-t.P1.Y) -
+		(t.P3.X-t.P1.X)*(t.P2.Y-t.P1.Y))
+	// TODO: also check area with an epsilon?
+	return area == 0.0
```
Better use `< 0.0001` or something, like your TODO suggests

rasky commented on src/tools/wallmap/objloader.go:
```diff
@@ -0,0 +1,194 @@
+			fmt.Println("Error in parseFace", r)
+			fmt.Printf("lineno: %+v | kw: +%v | data : %+v\n", lineno, kw, data)
+			fmt.Println("len(of.Vertices): ", len(of.Vertices))
+			os.Exit(1)
```
I'd exit with `panic(r)`, so at least you see the stack trace which is always useful

rasky commented on src/tools/wallmap/objloader.go:
```diff
@@ -0,0 +1,194 @@
+	for i, s := range data {
+		// we are only interested in the vertex index
+		sidx := strings.Split(s, "/")[0]
+		if _, err := fmt.Sscanf(sidx, "%d", &(vi[i])); err != nil {
```
`strconv.ParseInt` or `strconv.Atoi`

rasky commented on src/tools/wallmap/objloader.go:
```diff
@@ -0,0 +1,194 @@
+}
+
+// SetMin checks if a > b, then a will be set to the value of b.
+func SetMin(a *float64, b float64) {
```
Probably `UpdateMin` is clearer.

rasky commented on src/tools/wallmap/objloader.go:
```diff
@@ -0,0 +1,194 @@
+package main
```
You're being very liberal with usage of exported names. Even if it's just a file in the main package, you're making things harder if you evolve this into a small package later. I'd suggest to still think of it as a micro-package and use lowercase names for everything but main entry points.

arl replied:
Oh yes. Actually that's what I'd like to do, think of it as a micro
package. For example in the future more line parsing functions might be
added in order to support bigger set of the OBJ standard features.

Le 24 août 2016 01:25, "Giovanni Bajo" notifications@github.com a écrit :

> In src/tools/wallmap/objloader.go
> https://github.com/RookieGameDevs/surviveler/pull/3#discussion_r75970292
> :
> 
> > @@ -0,0 +1,194 @@
> > +package main
> 
> You're being very liberal with usage of exported names. Even if it's just
> a file in the main package, you're making things harder if you evolve this
> into a small package later. I'd suggest to still think of it as a
> micro-package and use lowercase names for everything but main entry points.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/RookieGameDevs/surviveler/pull/3/files/1f37e8cab422f8d46587682b13db9c63d44b6718#r75970292,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAdF6iafYww5BqaSHQM_ctUTYjZme4Itks5qi4FQgaJpZM4JrcmN
> .

rasky commented on src/tools/wallmap/objloader.go:
```diff
@@ -0,0 +1,194 @@
+	return nil
+}
+
+func (of *ObjFile) parseFace(lineno int, kw string, data []string) error {
```
As far as I can tell `lineno` is only useful for error reporting. I'd skip it: let this function returns whatever errors and you can wrap them in the callers like: `fmt.Errorf("line %d: error parsing face: %v", lineno, err)`

rasky commented on src/tools/wallmap/objloader.go:
```diff
@@ -0,0 +1,194 @@
+
+		lineno++
+	}
+	return &obj, nil
```
You need to check `scanner.Error` here

rasky commented on src/tools/wallmap/wallmap.go:
```diff
@@ -0,0 +1,95 @@
+	"os"
+
+	"github.com/fogleman/gg"
+	"golang.org/x/image/bmp"
```
any reason not to use `image/png` instead? It's part of the standard library

duploduplo replied:
@aurelien-rainone I'm duploduplo. Better to untag people that may be not interested in the project :D

Anyway I don't think that switching to png is a real problem on client side. It should be just a matter of changing the loader.

rasky commented:
As a general comment, I would have made the `obj` library parse the whole 3d meshes, computing the whole 3d bounding box, and do the ortho projection (in your case, drop the Z) directly during rendering (aka: just ignore the Z coordinate during rendering).

Can you attach a high-def rendering made with this tool?

arl commented on src/tools/wallmap/wallmap.go:
```diff
@@ -0,0 +1,95 @@
+	"os"
+
+	"github.com/fogleman/gg"
+	"golang.org/x/image/bmp"
```
We would have to sync with @duplo (@duploduplo ?!) for this as the
resulting image is also read client-side, in order to ease mouse picking,
eg show a red tower instead of green when the player is trying to build a
turret on a wall.

But it's true that a PNG being lossless and compressed, it would be better
to use PNG instead of BMP, thus  substantially decreasing the assets size,
and allowing to discard an external depencency on the Go side. Good catch 😉

Le 24 août 2016 01:30, "Giovanni Bajo" notifications@github.com a écrit :

> In src/tools/wallmap/wallmap.go
> https://github.com/RookieGameDevs/surviveler/pull/3#discussion_r75970851
> :
> 
> > @@ -0,0 +1,95 @@
> > +package main
> > +
> > +import (
> > -   "flag"
> > -   "fmt"
> > -   "math"
> > -   "os"
> >   +
> > -   "github.com/fogleman/gg"
> > -   "golang.org/x/image/bmp"
> 
> any reason not to use image/png instead? It's part of the standard library
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/RookieGameDevs/surviveler/pull/3/files/1f37e8cab422f8d46587682b13db9c63d44b6718#r75970851,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAdF6t0ytm1W8rCaUXXhzZZxUcRoR5eQks5qi4J9gaJpZM4JrcmN
> .

arl commented:
Oh I see, you mean remain in 3D all the way, and just discard the z coord
at rendering, thus making the use of 2D types and funcs useless in the OBJ
loader.
You are right, this would be better and would allow the OBJ loader
mini-package to remain focused on 3D.

For the attachment, better do it after the BMP to PNG switch, I don't know
about GitHub size limits for attachment. To give you an idea, I ran the
tool against our current walls mesh, with a factor 128, resulting BMP was
97M.

Le 24 août 2016 01:32, "Giovanni Bajo" notifications@github.com a écrit :

> As a general comment, I would have made the obj library parse the whole
> 3d meshes, computing the whole 3d bounding box, and do the ortho projection
> (in your case, drop the Z) directly during rendering (aka: just ignore the
> Z coordinate during rendering).
> 
> Can you attach a high-def rendering made with this tool?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/RookieGameDevs/surviveler/pull/3#issuecomment-241913252,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAdF6hQnaPDUGnFMYWt_rYpZd27wa_erks5qi4MdgaJpZM4JrcmN
> .

rasky commented:
Which makes me think that you should at least concert it to indexed (8bit) before exporting, irrespective of the format type.

## 

Giovanni Bajo
Develer S.r.l.
http://www.develer.com
Sent from mobile

> Il giorno 24 ago 2016, alle ore 08:34, Aurelien Rainone notifications@github.com ha scritto:
> 
> Oh I see, you mean remain in 3D all the way, and just discard the z coord
> at rendering, thus making the use of 2D types and funcs useless in the OBJ
> loader.
> You are right, this would be better and would allow the OBJ loader
> mini-package to remain focused on 3D.
> 
> For the attachment, better do it after the BMP to PNG switch, I don't know
> about GitHub size limits for attachment. To give you an idea, I ran the
> tool against our current walls mesh, with a factor 128, resulting BMP was
> 97M.
> 
> Le 24 août 2016 01:32, "Giovanni Bajo" notifications@github.com a écrit :
> 
> > As a general comment, I would have made the obj library parse the whole
> > 3d meshes, computing the whole 3d bounding box, and do the ortho projection
> > (in your case, drop the Z) directly during rendering (aka: just ignore the
> > Z coordinate during rendering).
> > 
> > Can you attach a high-def rendering made with this tool?
> > 
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > https://github.com/RookieGameDevs/surviveler/pull/3#issuecomment-241913252,
> > or mute the thread
> > https://github.com/notifications/unsubscribe-auth/AAdF6hQnaPDUGnFMYWt_rYpZd27wa_erks5qi4MdgaJpZM4JrcmN
> > .
> > 
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub, or mute the thread.

arl committed:
Remove debug print

arl committed:
Replace Point2/3 with versatile 4D Vertex (float array)

arl committed:
Add Vertex X, Y, Z, and W getters

arl committed:
Use Z as Y coordinate

Don't know if this a particularity of the obj file used, but the
axis i need to consider as the Y (for generating in 2D) is
actually the 3rd coord of the vertices definition (so it's Z)

duploduplo commented:
From what we decided last time, we just need 0/1 bitmaps. Using a 1bit depth contains all the pieces of information we need and should definitely decrease the output rendering.

I assume we are just drawing obstacles, and not inheriting the height of the obstacle from the wallmap. In that case we whould need a grayscale 8bit.

arl committed:
isDegenerate works on 3D triangles

arl committed:
Add AABB struct, use it to keep track of mesh aabb

arl committed:
Warn user in case the image has a negative offset

this offset is the translation from the 0,0 origin to the image origin.
Which by convention chosen in Surviveler, should be 0.

In fact we should probably warn if it's not 0

arl commented:
Indeed we just need 1 bit per pixel, walkable or non walkable.
PNG supports this (1 bit color depth)

Le 24 août 2016 09:31, "Lorenzo Berni" notifications@github.com a écrit :

> From what we decided last time, we just need 0/1 bitmaps. Using a 1bit
> depth contains all the pieces of information we need and should definitely
> decrease the output rendering.
> 
> I assume we are just drawing obstacles, and not inheriting the height of
> the obstacle from the wallmap. In that case we whould need a grayscale 8bit.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/RookieGameDevs/surviveler/pull/3#issuecomment-241980673,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAdF6okNECU4K2uULTq3mh1VGVPVkaksks5qi_NngaJpZM4JrcmN
> .

arl committed:
Rename SetMin/Max with updateMin/Max

arl commented:
Here is the current output, still in BMP format, but i convert it to PNG with an external tool in order to have an acceptable size:

``` bash
$ wallmap -scale 128 -bmp out.bmp ../../../data/map/walls.obj                                                                                                                                                                                                                                                                                                                
WARNING: the mesh 2D has negative offset(s), probably due to some negative coordinates: x:0|y:-1
mesh bounding box:  {0.000252 56.501854 -0.010531 3.000328 -0.001253 32.506012}
scaled bounding box:  {0.032256 7232.237312 -1.347968 384.041984 -0.160384 4160.769536}
image x, y offset:  0 -1
image width/height:  7233 4161
Rasterization completed
Creating out.bmp
```

Then, using imagemagick's `convert`:

``` bash
convert out.bmp out.png
```

``` bash
$ file out.* && ls -lh out.*                                                                                                                                                                                                                                                                                                                                                 
out.bmp: PC bitmap, Windows 3.x format, 7233 x 4161 x 24
out.png: PNG image data, 7233 x 4161, 8-bit grayscale, non-interlaced
-rw-rw-r-- 1 aurelien aurelien  87M Aug 25 10:39 out.bmp
-rw-rw-r-- 1 aurelien aurelien 154K Aug 25 10:35 out.png
```

![out](https://cloud.githubusercontent.com/assets/476650/17962197/0043643c-6ab0-11e6-8394-8f7516834bf4.png)

Of course the resulting output file (which ever format) will have a smaller bit depth, ideally 1 bit.

arl commented:
FYI, still need to address some of the points evoked in the review

arl committed:
Unexport ObjFile fields; add getters

iacopy commented:
Great image! 👍  (I can't add reaction... fail 👎 )

arl committed:
Add different vertex constructors

arl committed:
Consider determinant < 1e-5 as 0

arl committed:
Use strconv.Atoi instead of fmt.Sscanf

arl committed:
Add objloader_test.go

arl committed:
Check bufio.Scanner error

arl committed:
Cosmetics

arl committed:
Fix possible overflow

arl committed:
Cosmetics

arl committed:
Use errors.New if error string has no params

arl committed:
Adapt tool description in cli usage

arl committed:
Remove lineno arg from parseXXX functions

arl committed:
Let us panic in parseFace instead of recover()

arl committed:
Add converter.go, convert image color model

arl committed:
Generate png with black&white palette

arl committed:
Use int color components instead of float

arl committed:
Fix cli usage

arl commented:
I chose the png over the bmp format for various reasons, even if it needs modifying _a bit_ the reading code on client and server side. Resulting image is smaller and benefit of compression because there are just 2 colors.
@rasky Are you aware of a way to set the png image color format to 8 bit (or less) grayscale with the standard Go library? Can't seem to go further down than this, though it may actually not even be needed due to the very small size.

``` bash
$ file out.png
out.png: PNG image data, 7233 x 4161, 16-bit/color RGB, non-interlaced 
```

Here is the 24K PNG file generated on the map with a factor 32
![out](https://cloud.githubusercontent.com/assets/476650/17987979/8a50c372-6b23-11e6-9aae-6557ca195fe6.png)

arl committed:
Add ErrDegenerateTriangle

arl committed:
Line numbers technically start at 1

arl committed:
Cosmetics

arl committed:
Refine some string literals

arl committed:
Fix wrong char in format string
